mscore/scoretreeview.cpp

```diff
@@ -0,0 +1,100 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#include "scoretreeview.h"
+
+namespace Ms {
+
+//---------------------------------------------------------
+//   selectElementAndChildren
+//---------------------------------------------------------
+
+static void selectElementAndChildren(Element* el)
+{
+    el->score()->selection().add(el);
+    for (ScoreElement* child : (*el)) {
+        if (child) {
+            selectElementAndChildren(toElement(child));
+        }
+    }
+}
+
+//---------------------------------------------------------
+//   ScoreTreeView
+//---------------------------------------------------------
+
+ScoreTreeView::ScoreTreeView(QWidget* parent)
+    : QTreeView(parent)
+{
+}
+
+//---------------------------------------------------------
+//   ~ScoreTreeView
+//---------------------------------------------------------
+
+ScoreTreeView::~ScoreTreeView()
+{
+}
+
+//---------------------------------------------------------
+//   setModel
+//---------------------------------------------------------
+
+void ScoreTreeView::setModel(QAbstractItemModel* model)
+{
+    Q_ASSERT(model);  // ensure there is a model
+    Q_ASSERT(dynamic_cast<ScoreItemModel*>(model));  // ensure model is correct type
+    QTreeView::setModel(model);
+}
+
+//---------------------------------------------------------
+//   model
+//---------------------------------------------------------
+
+ScoreItemModel* ScoreTreeView::scoreItemModel() const
+{
+    return static_cast<ScoreItemModel*>(QTreeView::model());
+}
```

Comment:
It is safe and appropriate to just call this function `model()`. Callers that are aware of ScoreTreeView will get this function, while callers that treat this class as a QTreeView and try to call `model()` will be get `QTreeView::model()` instead.

```c++
QTreeView tv1 = ScoreTreeView(this);
auto model1 = tv1.model(); //  model1 has type QAbstractItemModel*

ScoreTreeView tv2 = ScoreTreeView(this);
auto model2 = tv2.model(); //  model2 has type ScoreItemModel*

```